gr-zeromq/lib/base_impl.cc

```diff
@@ -43,6 +47,42 @@ base_impl::base_impl(int type,
     if (major < 3) {
         d_timeout *= 1000;
     }
+
+    /* Set ZMQ_LINGER so socket won't infinitely block during teardown */
+#if USE_NEW_CPPZMQ_SET_GET
+    d_socket.set(zmq::sockopt::linger, LINGER_DEFAULT);
+#else
+    d_socket.setsockopt(ZMQ_LINGER, &LINGER_DEFAULT, sizeof(LINGER_DEFAULT));
+#endif
+
+    /* Set high watermark */
+    if (hwm >= 0) {
+#if USE_NEW_CPPZMQ_SET_GET
+        if (sink) {
+            d_socket.set(zmq::sockopt::sndhwm, hwm);
+        } else {
+            d_socket.set(zmq::sockopt::rcvhwm, hwm);
+        }
+#else
+#ifdef ZMQ_SNDHWM
+        if (sink) {
+            d_socket.setsockopt(ZMQ_SNDHWM, &hwm, sizeof(hwm));
+        } else {
+            d_socket.setsockopt(ZMQ_RCVHWM, &hwm, sizeof(hwm));
```

Comment:
This section of code is the problematic behavior pulled into the base class. Previously, it was correctly in the source/sink sub-classes, where each class did the appropriate thing. However, for the sake of a single place where you put the bind/connect logic, the base class interface and implementation has become polluted with what is more properly sub-class abstraction.

The challenge is in having polymorphic behavior in a constructor. I think the most straightforward solution would be to move the socket setup out of the ctor and into the `start()` method. Then, this base class could declare the following methods:

```c++
public:
    bool start() override;
protected:
    virtual void set_hwm(...) = 0;
    void link_socket();
```

and implement:
```c++
bool base_impl::start()
{
    set_hwm();
    link_to_socket();

    return gr::sync_block::start();
}

void link_socket()
{
    if (bind) {
         d_socket.bind(address);
     } else {
         d_socket.connect(address);
     }
}

```
This of course, comes at a cost of holding the HWM and the address as member state of the base class.